New Page Layout Picker: Convert new page layout modal to TypeScript

p-jackson opened the pull request:
#### Changes proposed in this Pull Request

Adds types to the `@automattic/page-template-modal` package, which should make it easier to work on in the future.

The code is based off the redesign done in #49860 (instead of `trunk`) because it was much easier to type due to the fact that a lot of code got deleted in #49860.

Most of the changes are just changing type annotations, however there are a few small logic changes to deal with undefined/null values the compiler caught.

Note that it's still using the `transpile` command from `@automattic/calypso-build` to build the package. This does seem to work though, so perhaps babel has been configured to work with TypeScript. Ping ~sgomes~ @scinos to make sure this is the correct thing to do.

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Build should succeed
* `yarn dev --sync` should still work, and the "new page layout modal" should still work correctly (sanity check, it shouldn't crash)

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

p-jackson committed:
Convert new page layout modal to TypeScript

sgomes commented:
> Note that it's still using the `transpile` command from `@automattic/calypso-build` to build the package. This does seem to work though, so perhaps babel has been configured to work with TypeScript. Ping @sgomes to make sure this is the correct thing to do.

As far as I know, `calypso-build` handles all TypeScript transpilation through Babel, yes, but I'm not the best person to ask. Perhaps @sirreal has more specific advice?

sirreal commented:
Yes, babel compiles TypeScript with calypso-build as far as I know. Babel always compiles TypeScript to build the Calypso application. We're using the TypeScript compiler (`tsc`) to build several packages for publish as well as to typecheck their soundness.

You could probably switch from `transpile` to `tsc` for compiling the package, several other packages do that. It's probably not too important.

p-jackson commented:
> Ping sgomes to make sure this is the correct thing to do.

Oops, sorry Sérgio, I meant Sergio @scinos

p-jackson committed:
Switch to building package using `tsc`

p-jackson committed:
Remove unused packages

scinos commented:
I suggest adding this new TS package to `./packages/tsconfig.json`. That way it will get compiled when the repo is installed, which will provide type annotations for IDE that supports it (eg: VSCode)

p-jackson committed:
Add package /packages/tsconfig.json so it gets built during `yarn install`

roo2 commented on packages/page-template-modal/package.json:
```diff
@@ -28,34 +28,26 @@
 		"src"
 	],
```
other typescript packages include the "types" root in `package.json`

```suggestion
	],
	"types": "dist/types",
```

p-jackson replied:
Ah yes, this is important! I missed that and would have seen it if I'd actually converted some of `starter-page-templates` to TS as well like you suggest.
I'll do that.

roo2 reviewed (commented):
I was trying to test whether the types are visible outside of the package, It would be easier to see it all working together if you actually convert the `*.js` code in `apps/editing-toolkit/editing-toolkit-plugin/starter-page-templates` as part of this PR, if you think that's a good idea?

but I think it needs the _types root_ in the `package.json`

The build of the ETK and the page-layout-modal are both working correctly ✅

p-jackson committed:
Add typings to editing-toolkit-plugin

p-jackson commented:
@roo2 I've added typings to the code in `starter-page-templates`, which was a good suggestion because 1. the typings weren't actually exported from the package and 2. I needed to export a couple more things explicitly to make it easier to use as a page: `LayoutDefinition` and `LayoutCategory`.

The diff for `starter-page-templates` isn't quite as clean to read unfortunately.
- The old component was built with `compose( withSelect( ... ), withDispatch( ... ) )` which is just harder to get the typings right for. So I converted it to a functional component that uses hooks to get at the state. The old code was assuming that intermediate props like `getMeta` and `postContentBlock` would just be ignored by `<PageTemplateModal>`. But TypeScript notices these things :)
- Adding types to the data store takes a bit of code

deBhal commented on packages/page-template-modal/tsconfig.json:
```diff
@@ -0,0 +1,32 @@
+		"module": "esnext",
+		"allowJs": true,
+		"checkJs": false,
+		"jsx": "react-jsx",
```
I notice we use `"jsx": "react"` nearly everywhere else in calypso. I see it's a [simple difference](https://www.typescriptlang.org/docs/handbook/jsx.html), but I'm not sure about the implications - is this intentional?

p-jackson replied:
Yes this was intentional, even though you're right most places are using `jsx=react`.

In a file that's using jsx syntax, the compiler expects there to be an `import * as React from 'react'` at the top of the file. And that's what most of the code in Calypso etc. does. If you're going to write React code then you have to import React first.

However not all the ETK code was doing this. In particular, the layout modal wasn't importing React at the top of its modules. I wanted to leave this as it is because I figured it was something to do with the fact that as a WP plugin this wasn't using Facebook's `react`, but using `@wordpress/element` (or whatever our React wrapper is called). However leaving it the way it was caused a TypeScript error: you can't use jsx syntax in a file unless you've imported React first.

**Unless** tsconfig has `jsx=react-jsx` (or `react-jsxdev`). This option was introduced in React 17 when the requirement to import React first was dropped. Using this option makes the TypeScript compiler happy with the way that the layout modal code was already dealing with jsx syntax (whatever way that was)

deBhal reviewed (commented):
LGTM!

No problems building and looks nice on wpcom and atomic sites:
![New_Page_‹_Julesaus_Test_Headstart_KO_—_WordPress_com](https://user-images.githubusercontent.com/5952255/112086774-9097db00-8bd8-11eb-80a3-c0bdaa332bd9.jpg)


Githubs moved file diffs are a bit annoying, aren't they? In case anyone else is looking at this:

`git diff `git merge-base HEAD trunk`:apps/editing-toolkit/editing-toolkit-plugin/starter-page-templates/page-template-plugin.js HEAD:apps/editing-toolkit/editing-toolkit-plugin/starter-page-templates/page-template-plugin.tsx`
`git diff `git merge-base HEAD trunk`:apps/editing-toolkit/editing-toolkit-plugin/starter-page-templates/store.js HEAD:apps/editing-toolkit/editing-toolkit-plugin/starter-page-templates/store.ts`

deBhal commented:
PS: That looks like a lot of work, thanks!
PPS: I saw those same e2e failures on completely unrelated code
